File: ring-android/app/src/main/java/cx/ring/account/AccountEditionFragment.java
/*
 *  Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Alexandre Lision <alexandre.lision@savoirfairelinux.com>
 *          Alexandre Savard <alexandre.savard@savoirfairelinux.com>
 *          Adrien Béraud <adrien.beraud@savoirfairelinux.com>
 *          Loïc Siret <loic.siret@savoirfairelinux.com>
 *          AmirHossein Naghshzan <amirhossein.naghshzan@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package cx.ring.account;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.FrameLayout;
import android.widget.LinearLayout;

import cx.ring.R;
import cx.ring.application.JamiApplication;
import cx.ring.client.HomeActivity;
import cx.ring.contactrequests.BlockListFragment;
import cx.ring.databinding.FragAccountSettingsBinding;
import cx.ring.fragments.AdvancedAccountFragment;
import cx.ring.fragments.GeneralAccountFragment;
import cx.ring.fragments.MediaPreferenceFragment;
import cx.ring.fragments.SecurityAccountFragment;
import cx.ring.interfaces.BackHandlerInterface;
import cx.ring.mvp.BaseSupportFragment;
import cx.ring.utils.DeviceUtils;

public class AccountEditionFragment extends BaseSupportFragment<AccountEditionPresenter> implements
        BackHandlerInterface,
        AccountEditionView,
        ViewTreeObserver.OnScrollChangedListener  {
    private static final String TAG = AccountEditionFragment.class.getSimpleName();

    public static final String ACCOUNT_ID_KEY = AccountEditionFragment.class.getCanonicalName() + "accountid";
    static final String ACCOUNT_HAS_PASSWORD_KEY = AccountEditionFragment.class.getCanonicalName() + "hasPassword";
    public static final String ACCOUNT_ID = TAG + "accountID";

    private static final int SCROLL_DIRECTION_UP = -1;

    private FragAccountSettingsBinding mBinding;

    private boolean mIsVisible;

    private String mAccountId;
    private boolean mAccountIsJami;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mBinding = FragAccountSettingsBinding.inflate(inflater, container, false);
        // dependency injection
        ((JamiApplication) getActivity().getApplication()).getInjectionComponent().inject(this);
        return mBinding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mBinding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        super.onViewCreated(view, savedInstanceState);

        mAccountId = getArguments().getString(ACCOUNT_ID);

        HomeActivity activity = (HomeActivity) getActivity();
        if (activity != null && DeviceUtils.isTablet(activity)) {
            activity.setTabletTitle(R.string.navigation_item_account);
        }

        mBinding.fragmentContainer.getViewTreeObserver().addOnScrollChangedListener(this);

        presenter.init(mAccountId);
    }

    @Override
    public void displaySummary(String accountId) {
        toggleView(accountId, true);
        FragmentManager fragmentManager = getChildFragmentManager();
        Fragment existingFragment = fragmentManager.findFragmentByTag(JamiAccountSummaryFragment.TAG);
        Bundle args = new Bundle();
        args.putString(ACCOUNT_ID_KEY, accountId);
        if (existingFragment == null) {
            JamiAccountSummaryFragment fragment = new JamiAccountSummaryFragment();
            fragment.setArguments(args);
            fragmentManager.beginTransaction()
                    .add(R.id.fragment_container, fragment, JamiAccountSummaryFragment.TAG)
                    .commit();
        } else {
            if (!existingFragment.isStateSaved())
                existingFragment.setArguments(args);
            ((JamiAccountSummaryFragment) existingFragment).setAccount(accountId);
        }
    }

    @Override
    public void displaySIPView(String accountId) {
        toggleView(accountId, false);
    }

    @Override
    public void initViewPager(String accountId, boolean isJami) {
        mBinding.pager.setOffscreenPageLimit(4);
        mBinding.slidingTabs.setupWithViewPager(mBinding.pager);
        mBinding.pager.setAdapter(new PreferencesPagerAdapter(getChildFragmentManager(), getActivity(), accountId, isJami));
        BlockListFragment existingFragment = (BlockListFragment) getChildFragmentManager().findFragmentByTag(BlockListFragment.TAG);
        if (existingFragment != null) {
            Bundle args = new Bundle();
            args.putString(ACCOUNT_ID_KEY, accountId);
            if (!existingFragment.isStateSaved())
                existingFragment.setArguments(args);
            existingFragment.setAccount(accountId);
        }
    }

    @Override
    public void goToBlackList(String accountId) {
        BlockListFragment blockListFragment = new BlockListFragment();
        Bundle args = new Bundle();
        args.putString(ACCOUNT_ID_KEY, accountId);
        blockListFragment.setArguments(args);
        getChildFragmentManager().beginTransaction()
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                .addToBackStack(BlockListFragment.TAG)
                .replace(R.id.fragment_container, blockListFragment, BlockListFragment.TAG)
                .commit();
        mBinding.slidingTabs.setVisibility(View.GONE);
        mBinding.pager.setVisibility(View.GONE);
        mBinding.fragmentContainer.setVisibility(View.VISIBLE);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, @NonNull MenuInflater inflater) {
        menu.clear();
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.bindView(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        setBackListenerEnabled(false);
    }

    public boolean onBackPressed() {
        if (mBinding == null)
            return false;
        mIsVisible = false;

        if (getActivity() instanceof HomeActivity)
            ((HomeActivity) getActivity()).setToolbarOutlineState(true);
        if (mBinding.fragmentContainer.getVisibility() != View.VISIBLE) {
            toggleView(mAccountId, mAccountIsJami);
            return true;
        }
        JamiAccountSummaryFragment summaryFragment = (JamiAccountSummaryFragment) getChildFragmentManager().findFragmentByTag(JamiAccountSummaryFragment.TAG);
        if (summaryFragment != null && summaryFragment.onBackPressed()) {
            return true;
        }
        return getChildFragmentManager().popBackStackImmediate();
    }

    private void toggleView(String accountId, boolean isJami) {
        mAccountId = accountId;
        mAccountIsJami = isJami;
        mBinding.slidingTabs.setVisibility(isJami? View.GONE : View.VISIBLE);
        mBinding.pager.setVisibility(isJami? View.GONE : View.VISIBLE);
        mBinding.fragmentContainer.setVisibility(isJami? View.VISIBLE : View.GONE);
        setBackListenerEnabled(isJami);
    }

    @Override
    public void exit() {
        Activity activity = getActivity();
        if (activity != null)
            activity.onBackPressed();
    }

    private void setBackListenerEnabled(boolean enable) {
        Activity activity = getActivity();
        if (activity instanceof HomeActivity)
            ((HomeActivity) activity).setAccountFragmentOnBackPressedListener(enable ? this : null);
    }

    private static class PreferencesPagerAdapter extends FragmentStatePagerAdapter {
        private final Context mContext;
        private final String accountId;
        private final boolean isJamiAccount;

        PreferencesPagerAdapter(FragmentManager fm, Context mContext, String accountId, boolean isJamiAccount) {
            super(fm);
            this.mContext = mContext;
            this.accountId = accountId;
            this.isJamiAccount = isJamiAccount;
        }

        @StringRes
        private static int getRingPanelTitle(int position) {
            switch (position) {
                case 0:
                    return R.string.account_preferences_basic_tab;
                case 1:
                    return R.string.account_preferences_media_tab;
                case 2:
                    return R.string.account_preferences_advanced_tab;
                default:
                    return -1;
            }
        }

        @StringRes
        private static int getSIPPanelTitle(int position) {
            switch (position) {
                case 0:
                    return R.string.account_preferences_basic_tab;
                case 1:
                    return R.string.account_preferences_media_tab;
                case 2:
                    return R.string.account_preferences_advanced_tab;
                case 3:
                    return R.string.account_preferences_security_tab;
                default:
                    return -1;
            }
        }

        @Override
        public int getCount() {
            return isJamiAccount ? 3 : 4;
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            return isJamiAccount ? getJamiPanel(position) : getSIPPanel(position);
        }

        @Override
        public CharSequence getPageTitle(int position) {
            int resId = isJamiAccount ? getRingPanelTitle(position) : getSIPPanelTitle(position);
            return mContext.getString(resId);
        }

        @NonNull
        private Fragment getJamiPanel(int position) {
            switch (position) {
                case 0:
                    return fragmentWithBundle(new GeneralAccountFragment());
                case 1:
                    return fragmentWithBundle(new MediaPreferenceFragment());
                case 2:
                    return fragmentWithBundle(new AdvancedAccountFragment());
                default:
                    throw new IllegalArgumentException();
            }
        }

        @NonNull
        private Fragment getSIPPanel(int position) {
            switch (position) {
                case 0:
                    return GeneralAccountFragment.newInstance(accountId);
                case 1:
                    return MediaPreferenceFragment.newInstance(accountId);
                case 2:
                    return fragmentWithBundle(new AdvancedAccountFragment());
                case 3:
                    return fragmentWithBundle(new SecurityAccountFragment());
                default:
                    throw new IllegalArgumentException();
            }
        }

        private Fragment fragmentWithBundle(Fragment result) {
            Bundle args = new Bundle();
            args.putString(ACCOUNT_ID_KEY, accountId);
            result.setArguments(args);
            return result;
        }
    }

    @Override
    public void onScrollChanged() {
        setupElevation();
    }

    private void setupElevation() {
        if (mBinding == null || !mIsVisible) {
            return;
        }
        Activity activity = getActivity();
        if (!(activity instanceof HomeActivity))
            return;
        LinearLayout ll = (LinearLayout) mBinding.pager.getChildAt(mBinding.pager.getCurrentItem());
        if (ll == null) return;
        RecyclerView rv = (RecyclerView)((FrameLayout) ll.getChildAt(0)).getChildAt(0);
        if (rv == null) return;
        HomeActivity homeActivity = (HomeActivity) activity;
        if (rv.canScrollVertically(SCROLL_DIRECTION_UP)) {
            mBinding.slidingTabs.setElevation(mBinding.slidingTabs.getResources().getDimension(R.dimen.toolbar_elevation));
            homeActivity.setToolbarElevation(true);
            homeActivity.setToolbarOutlineState(false);
        } else {
            mBinding.slidingTabs.setElevation(0);
            homeActivity.setToolbarElevation(false);
            homeActivity.setToolbarOutlineState(true);
        }
    }
}

File: ring-android/app/src/main/java/cx/ring/account/AccountEditionFragment.kt
/*
 *  Copyright (C) 2004-2022 Savoir-faire Linux Inc.
 *
 *  Author: Alexandre Lision <alexandre.lision@savoirfairelinux.com>
 *          Alexandre Savard <alexandre.savard@savoirfairelinux.com>
 *          Adrien Béraud <adrien.beraud@savoirfairelinux.com>
 *          Loïc Siret <loic.siret@savoirfairelinux.com>
 *          AmirHossein Naghshzan <amirhossein.naghshzan@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package cx.ring.account

import android.content.Context
import android.os.Bundle
import android.view.*
import android.view.ViewTreeObserver.OnScrollChangedListener
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.activity.OnBackPressedCallback
import androidx.annotation.StringRes
import androidx.fragment.app.*
import androidx.recyclerview.widget.RecyclerView
import cx.ring.R
import cx.ring.client.HomeActivity
import cx.ring.contactrequests.BlockListFragment
import cx.ring.databinding.FragAccountSettingsBinding
import cx.ring.fragments.AdvancedAccountFragment
import cx.ring.fragments.GeneralAccountFragment
import cx.ring.fragments.MediaPreferenceFragment
import cx.ring.fragments.SecurityAccountFragment
import cx.ring.mvp.BaseSupportFragment
import cx.ring.settings.pluginssettings.PluginsListSettingsFragment
import dagger.hilt.android.AndroidEntryPoint
import net.jami.account.AccountEditionPresenter
import net.jami.account.AccountEditionView

@AndroidEntryPoint
class AccountEditionFragment : BaseSupportFragment<AccountEditionPresenter, AccountEditionView>(),
    AccountEditionView, OnScrollChangedListener {
    private var mBinding: FragAccountSettingsBinding? = null
    private var mIsVisible = false
    private var mAccountId: String? = null
    private var mAccountIsJami = false

    private val callback = object : OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
            mIsVisible = false
            if (activity is HomeActivity) (activity as HomeActivity).setToolbarOutlineState(true)
            if (mBinding!!.fragmentContainer.visibility != View.VISIBLE) {
                toggleView(mAccountId, mAccountIsJami)
                return
            }
            val summaryFragment = childFragmentManager.findFragmentByTag(JamiAccountSummaryFragment.TAG) as JamiAccountSummaryFragment?
            if (!childFragmentManager.popBackStackImmediate()) {
                isEnabled = false
                requireActivity().onBackPressed()
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
        FragAccountSettingsBinding.inflate(inflater, container, false).apply { mBinding = this }.root

    override fun onDestroyView() {
        super.onDestroyView()
        mBinding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        setHasOptionsMenu(true)
        super.onViewCreated(view, savedInstanceState)
        mAccountId = requireArguments().getString(ACCOUNT_ID_KEY)
        (this.activity as HomeActivity?)?.setToolbarTitle(R.string.menu_item_account_settings)
        mBinding!!.fragmentContainer.viewTreeObserver.addOnScrollChangedListener(this)
        presenter.init(mAccountId!!)
    }

    override fun displaySummary(accountId: String) {
        toggleView(accountId, true)
        val fragmentManager = childFragmentManager
        val existingFragment = fragmentManager.findFragmentByTag(JamiAccountSummaryFragment.TAG)
        val args = Bundle().apply { putString(ACCOUNT_ID_KEY, accountId) }
        if (existingFragment == null) {
            val fragment = JamiAccountSummaryFragment().apply {
                arguments = args
            }
            fragmentManager.beginTransaction()
                .add(R.id.fragment_container, fragment, JamiAccountSummaryFragment.TAG)
                .commitAllowingStateLoss()
        } else {
            if (existingFragment is JamiAccountSummaryFragment) {
                if (!existingFragment.isStateSaved) existingFragment.arguments = args
                existingFragment.setAccount(accountId)
            }
        }
    }

    override fun displaySIPView(accountId: String) {
        toggleView(accountId, false)
    }

    override fun initViewPager(accountId: String, isJami: Boolean) {
        mBinding?.apply {
            pager.offscreenPageLimit = 4
            pager.adapter = PreferencesPagerAdapter(childFragmentManager, requireContext(), accountId, isJami)
            slidingTabs.setupWithViewPager(pager)
        }
        val existingFragment = childFragmentManager.findFragmentByTag(BlockListFragment.TAG) as BlockListFragment?
        if (existingFragment != null) {
            if (!existingFragment.isStateSaved)
                existingFragment.arguments = Bundle().apply { putString(ACCOUNT_ID_KEY, accountId) }
            existingFragment.setAccount(accountId)
        }
    }

    override fun goToBlackList(accountId: String) {
        val blockListFragment = BlockListFragment()
        blockListFragment.arguments = Bundle().apply { putString(ACCOUNT_ID_KEY, accountId) }
        childFragmentManager.beginTransaction()
            .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
            .addToBackStack(BlockListFragment.TAG)
            .replace(R.id.fragment_container, blockListFragment, BlockListFragment.TAG)
            .commit()
        mBinding?.apply {
            slidingTabs.visibility = View.GONE
            pager.visibility = View.GONE
            fragmentContainer.visibility = View.VISIBLE
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
    }

    override fun onResume() {
        super.onResume()
        presenter.bindView(this)
        requireActivity().onBackPressedDispatcher.addCallback(this, callback)
    }

    private fun toggleView(accountId: String?, isJami: Boolean) {
        mAccountId = accountId
        mAccountIsJami = isJami
        mBinding?.apply {
            slidingTabs.visibility = if (isJami) View.GONE else View.VISIBLE
            pager.visibility = if (isJami) View.GONE else View.VISIBLE
            fragmentContainer.visibility = if (isJami) View.VISIBLE else View.GONE
        }
    }

    override fun exit() {
        activity?.onBackPressed()
    }

    private class PreferencesPagerAdapter constructor(
        fm: FragmentManager,
        private val mContext: Context,
        private val accountId: String,
        private val isJamiAccount: Boolean
    ) : FragmentStatePagerAdapter(fm) {
        override fun getCount(): Int = if (isJamiAccount) 3 else 4

        override fun getItem(position: Int): Fragment {
            return if (isJamiAccount) getJamiPanel(position) else getSIPPanel(position)
        }

        override fun getPageTitle(position: Int): CharSequence = mContext.getString(if (isJamiAccount) getRingPanelTitle(position) else getSIPPanelTitle(position))

        private fun getJamiPanel(position: Int): Fragment {
            return when (position) {
                0 -> fragmentWithBundle(GeneralAccountFragment())
                1 -> fragmentWithBundle(MediaPreferenceFragment())
                2 -> fragmentWithBundle(AdvancedAccountFragment())
                3 -> fragmentWithBundle(PluginsListSettingsFragment())
                else -> throw IllegalArgumentException()
            }
        }

        private fun getSIPPanel(position: Int): Fragment {
            return when (position) {
                0 -> GeneralAccountFragment.newInstance(accountId)
                1 -> MediaPreferenceFragment.newInstance(accountId)
                2 -> fragmentWithBundle(AdvancedAccountFragment())
                3 -> fragmentWithBundle(SecurityAccountFragment())
                4 -> fragmentWithBundle(PluginsListSettingsFragment())
                else -> throw IllegalArgumentException()
            }
        }

        private fun fragmentWithBundle(result: Fragment): Fragment = result.apply {
            arguments = Bundle().apply { putString(ACCOUNT_ID_KEY, accountId) }
        }

        companion object {
            @StringRes
            private fun getRingPanelTitle(position: Int): Int {
                return when (position) {
                    0 -> R.string.account_preferences_basic_tab
                    1 -> R.string.account_preferences_media_tab
                    2 -> R.string.account_preferences_advanced_tab
                    3 -> R.string.account_preference_plugin_tab
                    else -> -1
                }
            }

            @StringRes
            private fun getSIPPanelTitle(position: Int): Int {
                return when (position) {
                    0 -> R.string.account_preferences_basic_tab
                    1 -> R.string.account_preferences_media_tab
                    2 -> R.string.account_preferences_advanced_tab
                    3 -> R.string.account_preferences_security_tab
                    4 -> R.string.account_preference_plugin_tab
                    else -> -1
                }
            }
        }
    }

    override fun onScrollChanged() {
        setupElevation()
    }

    private fun setupElevation() {
        val binding = mBinding ?: return
        if (!mIsVisible)
            return
        val activity: FragmentActivity = activity as? HomeActivity ?: return
        val ll = binding.pager.getChildAt(binding.pager.currentItem) as LinearLayout
        val rv = (ll.getChildAt(0) as FrameLayout).getChildAt(0) as RecyclerView
        val homeActivity = activity as HomeActivity
        if (rv.canScrollVertically(SCROLL_DIRECTION_UP)) {
            binding.slidingTabs.elevation = binding.slidingTabs.resources.getDimension(R.dimen.toolbar_elevation)
            homeActivity.setToolbarElevation(true)
            homeActivity.setToolbarOutlineState(false)
        } else {
            binding.slidingTabs.elevation = 0f
            homeActivity.setToolbarElevation(false)
            homeActivity.setToolbarOutlineState(true)
        }
    }

    companion object {
        private val TAG = AccountEditionFragment::class.simpleName
        val ACCOUNT_ID_KEY = AccountEditionFragment::class.qualifiedName + "accountId"
        val ACCOUNT_HAS_PASSWORD_KEY = AccountEditionFragment::class.qualifiedName + "hasPassword"
        private const val SCROLL_DIRECTION_UP = -1
    }
}